### PR TITLE
fix: array subscript doesn't have correct precedence 

### PR DESCRIPTION
**Pull request by CAJan93**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Issue description: https://github.com/risingwavelabs/risingwave/issues/5409

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by CAJan93** [comment]
The e2e-tests look promising. Can this really be this simple @xxchan ? Did I miss something? Do we assign a different precedence to operators depending on their relative position?

**Comment by xxchan** [comment]
> Can this really be this simple? 

I believe so 😄. FYI here'e the upstream's fix: https://github.com/sqlparser-rs/sqlparser-rs/pull/433/files

> Do we assign a different precedence to operators depending on their relative position?

I'm not sure I understand what you mean here correctly, but I guess not. Operator precedence will be constants.

`(array[1])[1] >= 1` works not because `[` has higher precedence when put at left, but actually because `parse_infix` has special handling of `[` (anyway it's not a binary op and waits for `]`), it calls `parse_infix` instead of recursing back to `parse_subexpr`.

File: src/sqlparser/src/parser.rs [hunk 1]
```diff
@@ -1201,7 +1201,7 @@ impl Parser {
             Token::Mul | Token::Div | Token::Mod | Token::Concat => Ok(40),
             Token::DoubleColon => Ok(50),
             Token::ExclamationMark => Ok(50),
-            Token::LBracket => Ok(10),
+            Token::LBracket => Ok(21),
```
Comment (xxchan) [inline src/sqlparser/src/parser.rs:1204]: This is not enough. `select 1*array[1][1]` will fail. I'm not sure but maybe using 50? (same as upstream's fix)

**Comment by CAJan93** [comment]
> I'm not sure I understand what you mean here correctly, but I guess not. Operator precedence will be constants.

I mean something like the following, very generic example: `1 - 2 * 3` is equivalent to `1 - (2 * 3)`, because `*` takes precedence over `-` when `-` is used as an infix operator. 

However in `-2 * 3` the `-` takes precedence over `*`. This is because `-` is used as a prefix instead of an infix.

**Comment by xxchan** [comment]
Got it, but I think this is not related to the problem of `[` since it has only one meaning as array index?

File: src/sqlparser/src/parser.rs (hunk 1 above)
Comment (CAJan93) [reply src/sqlparser/src/parser.rs]: Thank you very much for pointing out the upstream fix. I added the test as well @xxchan

**Comment by CAJan93** [comment]
> Got it, but I think this is not related to the problem of `[ since it has only one meaning as array index?

Probably, though I am not sure when considering the following queries: 

```SQL
select regexp_matches('foobarbequebazilbarfbonk', '(b[^b]+)(b[^b]+)', 'g');
create table t(x int[]); # Is about arrays, but not array access.
select 1::real || array[1]::numeric[]; # casting to array, not accessing it
```

**Comment by CAJan93** [comment]
Perfect. Let me know if the PR LGTY @xxchan

**Review by xxchan — APPROVED** [review]

**Pull request merged**